File: fabric-audit-agent/core/finding.js
/** @typedef {{level:'Critical'|'Warning'|'Info', reason:string}} Score */
/** @typedef {{what:string, where:string, when:string, why:string, impact:string, fix:string[], score:Score}} Finding */

const REQUIRED = ['what', 'where', 'when', 'why', 'impact', 'fix', 'score'];

/**
 * Build a validated 7-field finding.
 * @param {Finding} parts
 * @returns {Finding}
 */
export function createFinding(parts) {
  for (const k of REQUIRED) {
    if (parts[k] === undefined || parts[k] === null) {
      throw new Error(`createFinding: missing required field "${k}"`);
    }
  }
  if (!Array.isArray(parts.fix)) {
    throw new Error('createFinding: "fix" must be an array');
  }
  return {
    what: parts.what,
    where: parts.where,
    when: parts.when,
    why: parts.why,
    impact: parts.impact,
    fix: parts.fix,
    score: parts.score,
  };
}

/**
 * Wrap findings in the OS standard output envelope.
 * @param {{agentId:string, findings:Finding[], summary:string}} args
 */
export function wrapEnvelope({ agentId, findings, summary }) {
  return {
    success: true,
    agent_id: agentId,
    data: { findings },
    summary,
    timestamp: new Date().toISOString(),
  };
}

File: fabric-audit-agent-py/fabric_audit_agent/finding.py
"""Finding construction + output envelope. Faithful port of the Node ``core/finding.js``.

A finding has exactly 7 fields: what, where, when, why, impact, fix, score.
``when`` may be an empty string (only None/missing is rejected — matches the JS
``=== undefined || === null`` check, so falsy values like "" and 0 are allowed).
"""
from datetime import datetime, timezone

REQUIRED = ("what", "where", "when", "why", "impact", "fix", "score")


def create_finding(parts):
    """Build a validated 7-field finding (dict). Raises on a missing required field."""
    for k in REQUIRED:
        if parts.get(k) is None:
            raise ValueError(f'create_finding: missing required field "{k}"')
    if not isinstance(parts["fix"], list):
        raise TypeError('create_finding: "fix" must be a list')
    return {k: parts[k] for k in REQUIRED}


def wrap_envelope(*, agent_id, findings, summary):
    """Wrap findings in the standard output envelope."""
    return {
        "success": True,
        "agent_id": agent_id,
        "data": {"findings": findings},
        "summary": summary,
        "timestamp": datetime.now(timezone.utc).isoformat(),
    }
